File: Hashtable/2031. Count Subarrays With More Ones Than Zeros.cpp
class Solution {
public:
    int subarraysWithMoreZerosThanOnes(vector<int>& nums) {
        unordered_map<int, int> count;
        const int MOD = 1e9 + 7;
        count[0] = 1;
        int prev = 0;
        int res = 0;
        int total = 0;
        for (int num : nums) {
            if (num == 1) {
                total += 1;
                prev += count[total - 1];
            } else {
                total -= 1 ;
                prev -= count[total];
            }
            res = (res + prev) % MOD;
            count[total]++;

        }

        return res % MOD ;

    }
};

File: Hashtable/2031. Count Subarrays With More Ones Than Zeros.py
class Solution:
    def subarraysWithMoreZerosThanOnes(self, nums: List[int]) -> int:
        MOD = 10**9 + 7
        count = collections.defaultdict(int)
        count[0] = 1
        cur = 0
        prev = 0
        res = 0
        for num in nums:
            if num == 1:
                cur += 1
                prev += count[cur - 1]
            else:
                cur -= 1
                prev -= count[cur]
            res = (res + prev) % MOD
            count[cur] += 1

        return res